google/auth/aws.py

```diff
@@ -637,33 +661,54 @@ def __init__(
-        environment_id = credential_source.get("environment_id") or ""
-        self._aws_security_credentials_supplier = _DefaultAwsSecurityCredentialsSupplier(
-            credential_source
-        )
-        self._cred_verification_url = credential_source.get(
-            "regional_cred_verification_url"
-        )
+        if credential_source is None and aws_security_credentials_supplier is None:
+            raise exceptions.InvalidValue(
+                "A valid credential source or AWS security credentials supplier must be provided."
+            )
+        if (
+            credential_source is not None
+            and aws_security_credentials_supplier is not None
+        ):
+            raise exceptions.InvalidValue(
+                "AWS credential cannot have both a credential source and an AWS security credentials supplier."
+            )
 
-        # Get the environment ID. Currently, only one version supported (v1).
-        matches = re.match(r"^(aws)([\d]+)$", environment_id)
-        if matches:
-            env_id, env_version = matches.groups()
+        if aws_security_credentials_supplier:
+            self._aws_security_credentials_supplier = aws_security_credentials_supplier
+            self._cred_verification_url = (
+                _DEFAULT_AWS_REGIONAL_CREDENTIAL_VERIFICATION_URL
+            )
+            _LOGGER.info(
+                "AWS security credentials supplier is being used. Note that the external account credential does not cache the AWS security credentials so caching should be implemented in the supplier."
+            )
         else:
-            env_id, env_version = (None, None)
-
-        if env_id != "aws" or self._cred_verification_url is None:
-            raise exceptions.InvalidResource(
-                "No valid AWS 'credential_source' provided"
+            environment_id = credential_source.get("environment_id") or ""
+            self._aws_security_credentials_supplier = _DefaultAwsSecurityCredentialsSupplier(
+                credential_source
             )
-        elif int(env_version or "") != 1:
-            raise exceptions.InvalidValue(
-                "aws version '{}' is not supported in the current build.".format(
-                    env_version
-                )
+            self._cred_verification_url = credential_source.get(
+                "regional_cred_verification_url"
             )
+
+            # Get the environment ID. Currently, only one version supported (v1).
+            matches = re.match(r"^(aws)([\d]+)$", environment_id)
+            if matches:
+                env_id, env_version = matches.groups()
+            else:
+                env_id, env_version = (None, None)
+
+            if env_id != "aws" or self._cred_verification_url is None:
+                raise exceptions.InvalidResource(
+                    "No valid AWS 'credential_source' provided"
+                )
+            elif int(env_version or "") != 1:
```

Comment:
This still has the ValueError from the previous PR?

Reply:
This was part of the code that was just moved around that I was going to handle in b/328671918, not an actual change/regression made by this feature

Reply:
Got it - so there will be a PR after this

Reply:
Yeah, first PR was doing a refactor to support suppliers, this PR adds support for users to create credentials using their own suppliers, then there will be a follow up PR that fixes the older issues you pointed out in the previous PR. Trying to keep the individual PR's smaller is possible since this is such a big change